extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpAuthenticator.java

```diff
@@ -24,33 +29,40 @@
     @Inject
     IdentityProviderManager identityProviderManager;
 
-    final HttpAuthenticationMechanism mechanism;
+    final HttpAuthenticationMechanism[] mechanisms;
 
     public HttpAuthenticator() {
-        mechanism = null;
+        mechanisms = null;
     }
 
     @Inject
     public HttpAuthenticator(Instance<HttpAuthenticationMechanism> instance,
-            Instance<IdentityProvider<UsernamePasswordAuthenticationRequest>> usernamePassword) {
-        if (instance.isResolvable()) {
-            if (instance.isAmbiguous()) {
-                throw new IllegalStateException("Multiple HTTP authentication mechanisms are not implemented yet, discovered "
-                        + instance.stream().collect(Collectors.toList()));
+            Instance<IdentityProvider<?>> providers) {
+        List<HttpAuthenticationMechanism> mechanisms = new ArrayList<>();
+        for (HttpAuthenticationMechanism mechanism : instance) {
+            boolean notFound = false;
+            for (Class<? extends AuthenticationRequest> mechType : mechanism.getCredentialTypes()) {
+                boolean found = false;
+                for (IdentityProvider<?> i : providers) {
+                    if (i.getRequestType().equals(mechType)) {
```

Comment:
In case of those 3 extensions on the class path, can we get a case where `quarkus-oidc` will an identity provider with the other matching extensions ?
IMHO for a given credential type only a single authentication mechanism should be supported. So if the users add all 3 of those extensions plus some custom one for the same credential type then only one is selected which means they will need to choose what exactly is needed by adding only one of these extensions :-)

Reply:
It's not the credential type that is the issue, it is the mechanism that is used to provide it. FORM and BASIC are both username/password, however there is no confusion as one is a POST and the other is a header. If two mechanims are both using Bearer auth then that is a problem.

Reply:
@stuartwdouglas how the unexpected side-effects can be prevented then ?
Here is one case: `quarkus-smallrye-jwt` is configured to accept tokens as the cookies, `quarkus-oidc` `web-app` application is getting an `IdToken`, but  `quarkus-smallrye-jwt` `MpJwtValidator` is picked up to validate this token, which can either be unexpectedly accepted or rejected (say due to the different expiry limits which apply to OIDC access and id tokens). This is a user error I guess but it feels we need to tighten something here:

 May be do something as basic as restricting a number of `HttpAuthenticationMechanism` per credential type ?

 Or should users who need such combinations provide a custom mechanism which would delegate to either Form or Basic ? May be we can just ship an extension providing this composite mechanism. Or introduce `CompositeHttpAuthenticationMechanism` which if injected, takes priority over everything else and will return a set of `HttpAuthenticationMechanism`s. So if the user needs a composition, they'd only need to provide:
```
@ApplicationScoped
class FormBasicHttpAuthMechanism {
    Set<HttpAuthenticationMechanism> getMechanisms() {
         // return as set
         new FormHttpAuthenticationMechanism()
         new BasicHttpAuthenticationMechanism()
    }
}
```

etc